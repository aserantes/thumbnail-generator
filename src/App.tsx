import React from "react";
import { AppWrapper, Header } from "./components";
import { ImagePreview, ImageSelector, Thumbnails } from "./features";
import {
  getBypassValidation,
  getFileIsValid,
  getImagesReadyOnServer,
  getTheme,
} from "./store/selectors";
import { setFileToUploadData } from "./store/fileToUploadSlice";
import { useDispatch, useSelector } from "react-redux";
import { useWindowSize } from "./hooks";

export function App() {
  const theme = useSelector(getTheme);
  const fileIsValid = useSelector(getFileIsValid);
  const bypassValidation = useSelector(getBypassValidation);
  const imagesReadyOnServer = useSelector(getImagesReadyOnServer);
  const { height, width } = useWindowSize();
  const dispatch = useDispatch();

  const handleChange = () => {
    dispatch(setFileToUploadData({ bypassValidation: !bypassValidation }));
  };

  const currentComponent = () =>
    (imagesReadyOnServer && <Thumbnails />) ||
    (fileIsValid && <ImagePreview />) || <ImageSelector />;

  return (
    <AppWrapper
      backGroundColor={theme.colors.background}
      data-testid="App-wrapper"
      fontColor={theme.colors.text}
    >
      <div // make this a component Debug, along with state info and such
        style={{
          backgroundColor: "#000",
          flexDirection: "column",
          opacity: 0.6,
          color: "#fff",
          position: "absolute",
          left: "50%",
          transform: "translate(-50%, 0)",
          top: 0,
        }}
      >
        <div>{`w:${width} h:${height} - ${fileIsValid}`}</div>
        <div>fileIsValid: {fileIsValid ? "yes" : "no"}</div>
        <div>
          <input
            checked={bypassValidation}
            id="check"
            type="checkbox"
            onChange={handleChange}
          />
          <label htmlFor="check">disable FE validation</label>
        </div>
      </div>
      <Header />
      {currentComponent()}
    </AppWrapper>
  );
}
