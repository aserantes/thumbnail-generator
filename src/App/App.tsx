import React from "react";
import { Provider } from "react-redux";
import { ImagePreview } from "features/ImagePreview";
import { ImageSelector } from "features/ImageSelector";
import { Thumbnails } from "features/Thumbnails";

import store from "../store";
import styled from "@emotion/styled";

const Container = styled.div`
  background-color: gray;
  padding: 20p;
`;

export const App = () => {
  return (
    <Provider store={store}>
      <Container>
        <div>App Component</div>
        <ImagePreview />
        <ImageSelector />
        <Thumbnails />
      </Container>
    </Provider>
  );
};
