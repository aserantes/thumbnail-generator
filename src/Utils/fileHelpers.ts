import { MAX_FILE_SIZE, VALID_FILE_TYPES } from "Constants";

export const checkTypeValidity = (fileType: string): boolean => {
  console.log(`checkTypeValidityCalled with fileType: [${fileType}]`);

  return (
    fileType !== "" &&
    VALID_FILE_TYPES.some((validFileType) => {
      console.log(
        `iterating valid types [${VALID_FILE_TYPES}]against [${fileType}]`
      );
      console.log(
        `validFileType [${fileType}] includes [${validFileType}?] [${fileType.includes(
          validFileType
        )}]`
      );

      return fileType.includes(validFileType);
    })
  );
};

export const checkSizeValidity = (fileSize: number): boolean => {
  return fileSize !== 0 && fileSize < MAX_FILE_SIZE;
};
