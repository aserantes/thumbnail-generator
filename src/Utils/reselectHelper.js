import store from "Store";
import * as ReselectTools from "reselect-tools";
import * as selectors from "Store/selectors";

ReselectTools.registerSelectors(selectors);
ReselectTools.getStateWith(() => store.getState());
