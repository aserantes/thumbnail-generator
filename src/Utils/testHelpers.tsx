import React from "react";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { DefaultRootState } from "Store";
import { Provider } from "react-redux";
import { initialState as fileToUpload } from "Store/fileToUploadSlice";
import { render } from "@testing-library/react";

export const rootInitialState = {
  fileToUpload,
};

const mockStore = configureStore([thunk]);

export const renderWithRedux = (
  ui: JSX.Element,
  initialState: DefaultRootState = rootInitialState
) => {
  const store = mockStore(initialState);

  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    mockStore: store,
  };
};

// https://png-pixel.com/
export const greenPixelBase64 =
  "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M/wHwAEBgIApD5fRAAAAABJRU5ErkJggg==";

export const redPixelBase64 =
  "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8DwHwAFBQIAX8jx0gAAAABJRU5ErkJggg==";

/* export default function({ name = 'file.txt', size = 1024, type = 'plain/txt', lastModified = new Date() }) {
    const blob = new Blob(['a'.repeat(size)], { type });

    blob.lastModifiedDate = lastModified;

    return new File([blob], name);
}

const testFile = MockFile({
     type: 'image/png',
     size: 50000,
}); */

/*

interface MockFile {
    name: string;
    body: string;
    mimeType: string;
}

const createFileFromMockFile = (file: MockFile): File => {
    const blob = new Blob([file.body], { type: file.mimeType }) as any;
    blob['lastModifiedDate'] = new Date();
    blob['name'] = file.name;
    return blob as File;
};

const createMockFileList = (files: MockFile[]) => {
    const fileList: FileList = {
        length: files.length,
        item(index: number): File {
            return fileList[index];
        }
    };
    files.forEach((file, index) => fileList[index] = createFileFromMockFile(file));

    return fileList;
};

How to use

it('should convert a text file', async () => {
  const service: DocumentService = TestBed.get(DocumentService);

  const fileList = createMockFileList([
      {
          body: 'test',
          mimeType: 'text/plain',
          name: 'test.txt'
      }
  ]);

  const convertedFiles = await service.convertFilesToModelDocuments(fileList);

  expect(convertedFiles).toBe([
      {
          name: 'test.txt',
          body: 'dGVzdA==',
          mimeType: 'text/plain'
      }
  ]);
});

*/

export * from "@testing-library/react";

export { default as userEvent } from "@testing-library/user-event";
