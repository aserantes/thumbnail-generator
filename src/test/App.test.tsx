import React from "react";
import { App } from "App";
import { render, screen } from "Utils/testHelpers";

describe("<App />", () => {
  it("should render", () => {
    render(<App />);

    const component = screen.getByTestId("App-wrapper");

    expect(component).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    expect(render(<App />).asFragment()).toMatchSnapshot();
  });

  it("should show ImageSelector if no valid file is selected", () => {});

  it("should show ImagePreview if valid file is selected", () => {});

  it("should show Thumbails if fetch is fulfilled", () => {});

  /*    */
});

/*
https://redux.js.org/recipes/writing-tests
https://www.robinwieruch.de/react-testing-library
https://www.robinwieruch.de/react-connected-component-test
*/
