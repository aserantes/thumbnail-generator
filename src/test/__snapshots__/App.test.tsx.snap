// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders 1`] = `
<DocumentFragment>
  <div
    class="css-14jqeis"
  >
    <div>
      App Component
    </div>
    <div
      class="css-17ztkzs"
    >
      <button
        class="css-uxd16r"
        data-testid="button"
      >
        change test
      </button>
      <input
        class="css-qafjh4"
        data-testid="input"
        value=""
      />
      <div>
        ImagePreview Component
      </div>
      <div
        data-testid="div"
      >
        reduxStringInitialState
      </div>
    </div>
    <div
      class="css-1w5jlbf"
    >
      ImageSelector Component
    </div>
    <div
      class="css-119czso"
    >
      <div>
        Thumbnails Component
      </div>
    </div>
  </div>
</DocumentFragment>
`;
